File: src/15-challenge/solution.py
def draw_table(data: list[dict[str, str | int]]) -> str:
    table: list[str] = []

    keys: list[str] = list(data[0].keys())
    values: list[list[str]] = []
    paddings: list[int] = [len(key) for key in keys]

    for element in data:
        _values: list[str] = []

        i: int = 0
        for key in element:
            value: str = f"{element.get(key)}"
            paddings[i] = max(paddings[i], len(value))
            _values.append(value)
            i += 1

        values.append(_values)

    line_sep: str = "+"
    row_titles: str = "|"

    for i, padding in enumerate(iterable=paddings):
        title: str = keys[i][0].upper() + keys[i][1:]

        line_sep += "-" * (padding + 2) + "+"
        row_titles += f" {title.ljust(padding)} |"

    table.append(line_sep)
    table.append(row_titles)
    table.append(line_sep)

    for element in values:
        row: str = "|"
        for i, value in enumerate(iterable=element):
            padding: int = paddings[i]
            row += f" {value.ljust(padding)} |"

        table.append(row)

    table.append(line_sep)

    return "\n".join(table)

File: src/15-challenge/solution.js
/**
 * @param {Array<Object>} data
 * @returns {string}
 */
export default function drawTable(data) {
	/** @type {string[]}>} */
	const table = []

	/** @type {string[]}>} */
	const keys = Object.keys(data[0])

	/** @type {string[][]}>} */
	const values = []

	/** @type {number[]}>} */
	const paddings = keys.map((key) => key.length)

	for (const element of data) {
		/** @type {string[]}>} */
		const _values = []

		/** @type {number}>} */
		let i = 0

		for (const key in element) {
			/** @type {string}>} */
			const value = `${element[key]}`

			paddings[i] = Math.max(paddings[i], value.length)
			_values.push(value)
			i++
		}

		values.push(_values)
	}

	/** @type {string}>} */
	let lineSeparator = '+'

	/** @type {string}>} */
	let rowTitles = '|'

	for (let i = 0; i < paddings.length; i++) {
		/** @type {string}>} */
		const title = keys[i][0].toUpperCase() + keys[i].substring(1)

		/** @type {number}>} */
		const padding = paddings[i]

		lineSeparator += '-'.repeat(padding + 2) + '+'
		rowTitles += ` ${title.padEnd(padding)} |`
	}

	table.push(lineSeparator, rowTitles, lineSeparator)

	for (const element of values) {
		/** @type {string}>} */
		let row = '|'

		for (let i = 0; i < element.length; i++) {
			/** @type {string}>} */
			const value = element[i]

			/** @type {number}>} */
			const padding = paddings[i]

			row += ` ${value.padEnd(padding)} |`
		}

		table.push(row)
	}

	table.push(lineSeparator)

	return table.join('\n')
}
